wandb/sdk/lib/hashutil.py

```diff
@@ -61,17 +61,14 @@ def _md5_file_hasher(*paths: StrPath) -> _hashlib.HASH:
     md5_hash = _md5()
 
     # Note: We use str paths (instead of pathlib.Path objs) for minor perf improvements.
-    for path in sorted(map(str, paths)):
+    # Also, don't bother with empty files, mmap fails on them anyway.
+    nonempty_filepaths = (pth for pth in map(str, paths) if os.stat(pth).st_size)
```

Comment:
I think the original is preferable. This way requires you to `stat` every file before you `mmap` the first one (because `sorted()` evaluates the generator), increasing the distance between the `stat` and `mmap` calls if there are many files. In this case it's probably OK (we expect files not to change), but in general it's a pattern that should be avoided.

In fact, why not let `mmap` fail and catch the exception? Then you don't add any work on the happy path, and you do a similar amount of work for any file that's empty. And it's the right way to do IO in other cases (trying and handling errors rather than checking in an attempt to prevent them).

Reply:
> I think the original is preferable. This way requires you to `stat` every file before you `mmap` the first one (because `sorted()` evaluates the generator), increasing the distance between the `stat` and `mmap` calls if there are many files. In this case it's probably OK (we expect files not to change), but in general it's a pattern that should be avoided.

Good catch + good point, I'd overlooked this.  Originally I was hoping to save the overhead of `open()` but that's looking pretty negligible here, in most cases.  Anyway, I'll revert this.  Thanks!

Also, considering the number of edge cases we encounter around I/O (e.g. operations across filesystem boundaries), I'm happy taking the safer route

> In fact, why not let `mmap` fail and catch the exception? Then you don't add any work on the happy path, and you do a similar amount of work for any file that's empty. And it's the right way to do IO in other cases (trying and handling errors rather than checking in an attempt to prevent them).

I'd actually considered and tried this before in #8473 (not sure I ever committed the change).  Agreed this is the typical/recommended pattern for IO.

I think my main concern was that it's not obvious what `mmap` raises on encountering an empty file descriptor -- on macOS it's `ValueError`, but AFAIK this isn't documented anywhere in the [mmap docs](https://docs.python.org/3/library/mmap.html#mmap.mmap)

Reply:
> on macOS it's `ValueError`, but AFAIK this isn't documented anywhere in the [mmap docs](https://docs.python.org/3/library/mmap.html#mmap.mmap)

(...only in the cpython [source code](https://github.com/python/cpython/blob/986a4e1b6fcae7fe7a1d0a26aea446107dd58dd2/Modules/mmapmodule.c#L1573-L1591))

Reply:
Actually on reconsideration -- this is good enough for me.  We can always change the `except ...:` clause on the off-chance there are breaking/platform-specific changes to this behavior

Reply:
done